File: backend/app/schemas/curriculum.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class DifficultyLevel(str, Enum):
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class ExerciseType(str, Enum):
    MULTIPLE_CHOICE = "multiple_choice"
    NUMERIC = "numeric"
    ALGEBRA = "algebra"


# ============================================================================
# Grade Schemas
# ============================================================================

class GradeBase(BaseModel):
    grade_number: int = Field(..., ge=5, le=7, description="Grade number (5-7)")


class GradeCreate(GradeBase):
    pass


class Grade(GradeBase):
    id: int
    
    class Config:
        from_attributes = True


# ============================================================================
# Topic Schemas
# ============================================================================

class TopicBase(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    grade_id: int


class TopicCreate(TopicBase):
    pass


class Topic(TopicBase):
    id: int
    
    class Config:
        from_attributes = True


# ============================================================================
# Lesson Schemas
# ============================================================================

class LessonBase(BaseModel):
    title: str = Field(..., max_length=255)
    content: Optional[str] = None
    topic_id: int


class LessonCreate(LessonBase):
    pass


class Lesson(LessonBase):
    id: int
    
    class Config:
        from_attributes = True


# ============================================================================
# Exercise Schemas
# ============================================================================

class ExerciseBase(BaseModel):
    question: str
    answer: str
    solution: Optional[str] = None
    difficulty: DifficultyLevel = DifficultyLevel.MEDIUM
    exercise_type: ExerciseType = ExerciseType.NUMERIC
    lesson_id: int


class ExerciseCreate(ExerciseBase):
    pass


class Exercise(ExerciseBase):
    id: int
    
    class Config:
        from_attributes = True


# Exercise without answer (for students)
class ExercisePublic(BaseModel):
    id: int
    question: str
    difficulty: DifficultyLevel
    exercise_type: ExerciseType
    lesson_id: int
    
    class Config:
        from_attributes = True


# ============================================================================
# ExerciseAttempt Schemas
# ============================================================================

class ExerciseAttemptBase(BaseModel):
    submitted_answer: str
    exercise_id: int


class ExerciseAttemptCreate(BaseModel):
    answer: str  # Simplified for API


class ExerciseAttempt(BaseModel):
    id: int
    user_id: int
    exercise_id: int
    submitted_answer: str
    is_correct: bool
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============================================================================
# Response Schemas
# ============================================================================

class ExerciseSubmissionResponse(BaseModel):
    correct: bool
    solution: str
    submitted_answer: str
    correct_answer: Optional[str] = None
    xp_gained: int = 0
    leveled_up: bool = False
    new_level: int = 1


class GradeWithTopics(Grade):
    topics: List[Topic] = []


class TopicWithLessons(Topic):
    lessons: List[Lesson] = []


class LessonWithExercises(Lesson):
    exercises: List[ExercisePublic] = []


class GeneratedTheoryResponse(BaseModel):
    lesson_id: int
    title: str
    content: str


class VideoSearchQueriesResponse(BaseModel):
    lesson_id: int
    queries: List[str]


class GeneratedExampleItem(BaseModel):
    difficulty: str
    problem: str
    solution: str


class GeneratedExamplesResponse(BaseModel):
    lesson_id: int
    title: str
    examples: List[GeneratedExampleItem]

File: frontend/src/services/curriculum.ts
import apiClient from './api';
import { trackEvent } from './analytics';

// ============================================================================
// Curriculum API Functions
// ============================================================================

export interface Grade {
  id: number;
  grade_number: number;
}

export interface Topic {
  id: number;
  title: string;
  description?: string;
  grade_id: number;
}

export interface Lesson {
  id: number;
  title: string;
  content?: string;
  topic_id: number;
}

export interface Exercise {
  id: number;
  question: string;
  difficulty: 'easy' | 'medium' | 'hard';
  exercise_type: 'multiple_choice' | 'numeric' | 'algebra';
  lesson_id: number;
}

export interface ExerciseSubmissionResponse {
  correct: boolean;
  solution: string;
  submitted_answer: string;
  correct_answer?: string;
  xp_gained: number;
  leveled_up: boolean;
  new_level: number;
}

export interface GeneratedTheory {
  lesson_id: number;
  title: string;
  content: string;
}

export interface VideoSearchQueries {
  lesson_id: number;
  queries: string[];
}

export interface GeneratedExampleItem {
  difficulty: 'easy' | 'medium' | 'hard';
  problem: string;
  solution: string;
}

export interface GeneratedExamples {
  lesson_id: number;
  title: string;
  examples: GeneratedExampleItem[];
}

/**
 * Get all available grades
 */
export const getGrades = async (): Promise<Grade[]> => {
  const response = await apiClient.get('/curriculum/grades');
  return response.data;
};

/**
 * Get all topics for a specific grade
 */
export const getTopics = async (gradeId: number): Promise<Topic[]> => {
  const response = await apiClient.get(`/curriculum/grades/${gradeId}/topics`);
  return response.data;
};

/**
 * Get all lessons for a specific topic
 */
export const getLessons = async (topicId: number): Promise<Lesson[]> => {
  const response = await apiClient.get(`/curriculum/topics/${topicId}/lessons`);
  return response.data;
};

/**
 * Get all exercises for a specific lesson
 */
export const getExercises = async (lessonId: number): Promise<Exercise[]> => {
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/exercises`);
  return response.data;
};

/**
 * Get a specific lesson with details
 */

export const getLesson = async (lessonId: number): Promise<Lesson> => {
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}`, {
    timeout: 60000,
  });
  return response.data;
};

export interface GeneratedContentStatus {
  lesson_id: number;
  cached_levels: string[];
}

/**
 * Get which detail levels are already cached in the database for a lesson
 */
export const getContentStatus = async (lessonId: number): Promise<GeneratedContentStatus> => {
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/content-status`);
  return response.data;
};

/**
 * Generate theory content for a lesson using AI
 */

export const getGeneratedTheory = async (
  lessonId: number,
  detailLevel: 'concise' | 'standard' | 'detailed' = 'standard',
): Promise<GeneratedTheory> => {
  trackEvent('ai_request', {
    endpoint: '/curriculum/lessons/{lessonId}/generated-theory',
    lesson_id: lessonId,
    detail_level: detailLevel,
  });
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/generated-theory`, {
    params: { detail_level: detailLevel },
    timeout: 60000,
  });
  return response.data;
};

/**
 * Get AI-rephrased YouTube search queries for a lesson
 */
export const getVideoSearchQueries = async (lessonId: number): Promise<VideoSearchQueries> => {
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/video-search-queries`);
  return response.data;
};

/**
 * Generate example problems for a lesson using AI
 */

export const getGeneratedExamples = async (lessonId: number): Promise<GeneratedExamples> => {
  trackEvent('ai_request', {
    endpoint: '/curriculum/lessons/{lessonId}/generated-examples',
    lesson_id: lessonId,
  });
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/generated-examples`, {
    timeout: 60000,
  });
  return response.data;
};

/**
 * Get a specific topic
 */
export const getTopic = async (topicId: number): Promise<Topic> => {
  const response = await apiClient.get(`/curriculum/topics/${topicId}`);
  return response.data;
};

/**
 * Get AI-generated exercises for a lesson (generates & caches on first call)
 */
export const getAIExercises = async (lessonId: number, regenerate = false): Promise<Exercise[]> => {
  trackEvent('ai_request', {
    endpoint: '/curriculum/lessons/{lessonId}/ai-exercises',
    lesson_id: lessonId,
    regenerate,
  });
  const response = await apiClient.get(`/curriculum/lessons/${lessonId}/ai-exercises`, {
    params: regenerate ? { regenerate: true } : {},
    timeout: 60000,
  });
  return response.data;
};

/**
 * Submit an answer to an exercise
 */
export const submitAnswer = async (
  exerciseId: number,
  answer: string
): Promise<ExerciseSubmissionResponse> => {
  const response = await apiClient.post(`/exercises/${exerciseId}/submit`, {
    answer: answer,
  });
  return response.data;
};
